packages/react-router/src/route.ts

```diff
@@ -122,39 +122,64 @@ export interface FullSearchSchemaOption<TFullSearchSchema> {
   search: TFullSearchSchema
 }
 
+export type RouteContextFn<TFullSearchSchema, TAllParams, TAllContext> = (
+  ctx: ContextOptions<TFullSearchSchema, TAllParams, TAllContext>,
+) => any
+
 export type FileBaseRouteOptions<
   TPath extends string = string,
   TSearchSchemaInput = Record<string, unknown>,
   TSearchSchema = {},
   TFullSearchSchema = TSearchSchema,
   TParams = {},
   TAllParams = {},
-  TRouteContextReturn = RouteContext,
-  TParentAllContext = AnyContext,
-  TAllContext = AnyContext,
   TLoaderDeps extends Record<string, any> = {},
   TLoaderDataReturn = {},
-> = {
+  TParentAllContext = {},
+  TRouteContextFn = AnyContext,
+  TBeforeLoadFn = AnyContext,
+> = ParamsOptions<TPath, TParams> & {
   validateSearch?:
-    | ((input: TSearchSchemaInput) => TSearchSchema)
     | { parse: (input: TSearchSchemaInput) => TSearchSchema }
+    | ((input: TSearchSchemaInput) => TSearchSchema)
+
   shouldReload?:
     | boolean
     | ((
-        match: LoaderFnContext<TAllParams, TFullSearchSchema, TAllContext>,
+        match: LoaderFnContext<
+          TAllParams,
+          TFullSearchSchema,
+          ResolveAllContext<TParentAllContext, TRouteContextFn, TBeforeLoadFn>
+        >,
       ) => any)
+
+  routeContext?: Constrain<
```

Comment:
Limitations of default type parameters and constraints leads me to use this pattern which I've made into a utility type. Long story short, it easy to get stale types which are the default types/constraints and not the inferred types. 

Instead `Constrain` is a conditional type which checks if a type is assignable to another type, if it is then all good, we can infer the type, otherwise we use the target type to force a type check error